deploy django app on okteto (#38)
okteto/blog-proposals: deploying-a-django-app-on-okteto.md

# Deploying a Django App on Okteto

## Introduction

In this tutorial, We will building and deploying a Django Role based Authorization Service API(RBAS) on Okteto.

> An RBAS application, is a system with checks to give access to users based on their roles in the system. 

Reviewer: ```suggestion
An RBAC application, is a system with checks to give access to users based on their roles in the system. 
```

Reviewer: could you add a sentence or two listing why this type of applications are important, or giving an example of what using RBAC allows?

What is Okteto ?

> Okteto is a cloud platform that enables developers develop and deploy cloud like Docker and Kubernetes applications. 

Reviewer: ```suggestion
Okteto is a free developer platform powered by Kubernetes that enables anyone to develop and deploy cloud-native applications.
```

What is Django?

> Django is one of the web framework of python for building APIs. Django is a framework that comes with out of the box functionality used in developing web apps and API, making your development faster and easier.

Reviewer: ```suggestion
Django is a high-level Python Web framework very well suited for building APIs. Django comes with a lot of out of the box functionality, making your development faster and easier.
```

Reviewer: could you link to the main Django page?

> ## Prerequisites 

Reviewer: Please include the version of python that should be installed.

- Knowledge of building APIs
- Basic Knowledge of Django
- Basic Knowledge of Cloud Infrastructure 

## Scaffold Project

To start building, we will first create and activate a virtual environment for our project:

    $ python3 -m venv venv
    $ source venv/bin/activate

Next, we install Django:

    $ pip install django

> Running `django-admin`. 

Reviewer: I think it's clearer if you merge the two paragraphs that talk about django-admin together like this:

```suggestion
With Django installed, go ahead and run the  `django-admin` command.  This command is provided by Django to help bootstrap your project, creating essential directories and files to get started.
```

    $ django-admin startproject simple_rbas

`django-admin` is a command provided by Django to help bootstrap your project, creating essential directories and files to get started.

Navigate into the new project created;

    $ cd simple_rbas

Create `requirements.txt` file in the root folder and add the following packages:

    django
    djangorestframework
    django-environ
    psycopg2-binary

>     gunicorn

Reviewer: could add a second program explaining why we need this libraries?

Install the packages using the `pip` command:

>     pip install -r requirements.txt 

Reviewer: could you add a final paragraph for this section explaining what did we accomplish so far?

## Development

In this section, we will be building up our model, serializers, permissions for our views, and routing to our views.

We will also be setting Docker and Okteto for our project.

### Setting Up Project

To get started with building our app, we’ll create an app inside our project.  Run this command inside your project directory;

    python manage.py startapp users

 
 This command creates a new directory named `users` in your project directory, looking like this:
 

    users
      - migrations
        -- __init__.py
      -- __init__.py
      -- admin.py
      -- apps.py
      -- models.py
      -- tests.py
      -- views.py

> For our authentication, we will be using using `http-only cookies` and not `localStorage`, to set our authentication.  This is to advised to reduce the risk of security issues.

Reviewer: ```suggestion
For our authentication, we will be using using `http-only cookies` and not `localStorage`, to set our authentication.  It is recommended to use ` http-only-cookies` to reduce the risk of security issues.
```

> We will be creating an authentication middleware and  for the project;

Reviewer: ```suggestion
We will be creating an authentication middleware  for the project;
```

Reviewer: Could you add a small paragraph before explaining why we are using middleware, how that's a common pattern for Django apps, or something like that?

                return None
            try:
                access_token = authorization_cookie.split(" ")[1]
                payload = decrypt_jwt_token(access_token)
            except JWTError:
                return None
            if not payload:
                return None
            user = User.objects.filter(id=payload["user_id"]).first()
            if user is None:
                raise exceptions.AuthenticationFailed("User not found")
            if not user.is_active:
                raise exceptions.AuthenticationFailed("User is inactive")
            self.enforce_csrf(request)
            return user, None
        def enforce_csrf(self, request):
            """
            Enforces CSRF Validation
            :param request:
            :return:
            """
            return
    

In the code above, `generate_jwt_token()` is a function that will be used in generate a `jwt_token`, once the user has been authenticated. 
`decrypt_jwt_token()` is used in our authentication middleware, `SafeJWTAuthentication` to get the `user` data, when a user accesses an endpoint. 

Next, we will be adding our authentication middleware to our configuration.

> In your `simple_rbas/settings` in your project directory add the following changes:

Reviewer: is it ` simple_rbas`or ` simple_rbac`  ?

    from pathlib import Path
    
    import environ
    
    ROOT_DIR = (
        environ.Path(__file__) - 2
        #image-repo/image_repo/settings.py
    )
    # Build paths inside the project like this: BASE_DIR / 'subdir'.
    BASE_DIR = Path(__file__).resolve().parent.parent
    env = environ.Env(
        # set casting, default value
        DEBUG=(bool, False),
    )
    # Build paths inside the project like this: BASE_DIR / 'subdir'.
    BASE_DIR = Path(__file__).resolve().parent.parent
    READ_DOT_ENV_FILE = env.bool("DJANGO_READ_DOT_ENV_FILE", default=True)
    if READ_DOT_ENV_FILE:
        # OS environment variables take precedence over variables from .env
        env.read_env(str(ROOT_DIR.path(".env")))

Here we are specifying the directory the `.env` file we will be creating.

Create a `.env`  file in your root directory:
`.env`

    SECRET_KEY=fake-secret-key
    DEBUG=True
    COOKIE_SECURE=False

Replace `fake-secret-key` with the `SECRET_KEY` value in settings.py.

> Our `env` keys will only be used for local development.

> Your secret key should be the `SECRET_KEY` string in the `settings.py` and change the `SECRET_KEY` and `DEBUG` variable in `settings.py`;

Reviewer: I'm confused about this step. what are you supposed to do here?

    # SECURITY WARNING: keep the secret key used in production secret!
    SECRET_KEY = env.str("SECRET_KEY") 
    # SECURITY WARNING: don't run with debug turned on in production!
    DEBUG = env.bool("DEBUG")
    
    ALLOWED_HOSTS = ['*']

Add the `users` app to the `INSTALLED_APPS` list:

    INSTALLED_APPS = [
        'django.contrib.admin',
        'django.contrib.auth',
        'django.contrib.contenttypes',
        'django.contrib.sessions',
        'django.contrib.messages',
        'django.contrib.staticfiles',
        'users'
    ]

Change the `DATABASE` variable:

    DATABASES = {"default": env.db("DATABASE_URL", default= "sqlite:////" + str(BASE_DIR) + '/db.sqlite3')}

> Finally, add the following lines at the end of the file;

Reviewer: can you explain what is this for?

          - DATABASE_URL=postgresql://pguser:pgpass@notes-db:5432/pgdb
          - COOKIE_SECURE=false
          - DEBUG=false

The `okteto-stack.yml` file is similar to the `docker-compose.yml` file. In the stack file, we are configuring two services: 

1. `rbas-db` for our database 
2.  `simple-rbas` our web API.  

Our services are linked together through the `DATABASE_URL` and both services are exposed to a certain port, `5432` and `8000`. 

The `rbas-db` is created from the `postgres11` image, and `simple-rbas` is based on the image created by our `Dockerfile`. 

### Initializing Models and Permissions

In `users/models.py`, add the following lines:

    from django.contrib.auth.models import AbstractUser
    from django.db import models
    
    
    class User(AbstractUser):
        OWNER = 0
        INVESTOR = 1
        ADMIN = 2
        WRITE = 3
        READ_ONLY = 4
        LEVEL_CHOICES = (
            (OWNER, "Owner"),
            (INVESTOR, "Investor"),
            (ADMIN, "Admin"),
            (WRITE, "Read/Write"),
            (READ_ONLY, "Read Only"),
        )
        level = models.SmallIntegerField(choices=LEVEL_CHOICES, db_index=True, default=READ_ONLY)

> Authorization on the API will be based user’s level. Only `Owner` users can have access to all views. 

Reviewer: ```suggestion
Authorization on the API will be based on the user’s level property. Only `Owner` users can have access to all views. 
```

`OnlyOwnerUser` is only available to owners.
These are through the permission classes attached to them. 

Add the following to your `urls.py`:
Here were are assigning the views we created to routes:

    from users.views import *
    
    urlpatterns = [
        path('admin/', admin.site.urls),
    ]
    
    urlpatterns += [
        path('user-auth/', LoginView.as_view({"post": "login"}), name = "user-login"),
        path('user-signup/', SignUpView.as_view({"post": "create"}), name = "user-signup"),
        path('all-user/', AllUser.as_view({"post": "get"}), name = "all-users"),
        path('authenticated-user/', OnlyAuthenticatedUser.as_view({"post": "get"}), name = "authenticated-user"),
        path('only-staff/', OnlyStaffOwnerUser.as_view({"post": "get"}), name = "only-staff"),
        path('only-admin/', OnlyAdminStaffOwnerUser.as_view({"post": "get"}), name = "only-admin"),
        path('only-investor-owner/', OnlyInvestorAndOwnerUser.as_view({"post": "get"}), name = "only-investor-owner"),
        path('only-owner/', OnlyOwnerUser.as_view({"post": "get"}), name = "only-owner"),
    ]
## Deploying to Okteto

> Now we will  be deploying our API and to Docker and Okteto Cloud.

Reviewer: ```suggestion
Now we will  be deploying our API and to  Okteto Cloud.
```

> Get an account on Okteto and run this command;

Reviewer: ```suggestion
Sign up for a [free Okteto Cloud account](https://cloud.okteto.com)  and then run this command;
```

> Get an account on Okteto and run this command;

Reviewer: can you first tell the reader to install the Okteto CLI, and point them to the instructions on https://okteto.com/docs/getting-started/installation

>     $ okteto namespace

Reviewer: could you add a sentence on what this command does?

Next we try building and deploying to Okteto:

    $ okteto stack deploy --build

> This creates our Docker image from our `Dockerfile`, and builds the services in the `okteto-stack.yml` file. 

Reviewer: ```suggestion
This command will be our API's container image using our `Dockerfile`, and then deploy the services in Okteto Cloud.
```

> Next, go to the Okteto Dashboard settings, and the secrets tab and add your secret key(gotten from your django settings or .env) for your Docker image:

Reviewer: This step doesn't work, since the secrets are only available during the deployment of a pipeline. (https://okteto.com/docs/cloud/secrets). I think for your post it's better to ignore this step, and instead use the `env_file` property of Okteto Stacks (https://okteto.com/docs/reference/stacks#env_file-string-optional).

![](https://paper-attachments.dropbox.com/s_63C04585E62AD95AAE546582E54FC11FBBFCB4DF4EB5E31E04B3F30729CE7718_1616271618177_okteto-dashboard.png)

Now to run migrations and set up your app finally:

    $ okteto up

Select the yes (y) option to create a `okteto.yml`. 

This sets up a development container for you.  `exit` and edit line 4:

    image: okteto.dev/simple-rbas:latest

Now run `okteto up` again

Run:

    $ python manage.py makemigrations

>     $ python manage.py migrate

Reviewer: I think it's better to put this on a script, and run it as part of the docker file, so that, when you run "okteto stack deploy --build", you have a functional application without any extra steps.

Reviewer: https://github.com/okteto/math/blob/master/run_web.sh is an example of how to do that.
